### PR TITLE
Allow FTP sync for non-BNN formats

### PR DESCRIPTION
**Pull request by JuliusR**
See http://foodsoft.51229.x6.nabble.com/Use-sharedlists-with-non-BNN-format-tt2309.html

TODO:
- [x] Update `README.md`
- [x] Rebase (and combine the two database migrations) if that is a good thing

### CODE REVIEW

File: app/models/supplier.rb [hunk 1]
```diff
@@ -6,26 +6,35 @@ class Supplier < ActiveRecord::Base
...
     Rails.root.join("supplier_assets", "bnn_files", id.to_s)
   end
 
   def mail_path
     Rails.root.join("supplier_assets", "mail_files", id.to_s)
   end
 
+  def ftp_type_string
+    FTP_TYPES.key(ftp_type).to_s
+  end
```
Comment (wvengen) [inline app/models/supplier.rb]: Perhaps I'd rather make this something like:
```ruby
def ftp_type
  FTP_TYPES.key(super)
end
```

You can still access the `attributes` if you really need the integer.

Or put the string in the database (that's what I usually do, it makes it more obvious what it is, and there is not a need for high performance here).

File: app/models/supplier.rb [hunk 2]
```diff
@@ -6,26 +6,35 @@ class Supplier < ActiveRecord::Base
...
 
+  # @rails4: enum ftp_type: {bnn: 0, foodsoft: 1}
+  FTP_TYPES = {bnn: 0, foodsoft: 1}
+
   EMAIL_RE = /^([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})$/i.freeze
 
   validates :name, :address, :phone, presence: true
-  validates :bnn_host, :bnn_user, :bnn_password, :bnn_sync, presence: true, if: :bnn_sync?
+  validates :ftp_host, :ftp_user, :ftp_password, :ftp_sync, presence: true, if: :ftp_sync?
+  validates :ftp_type, presence: true, inclusion: {in: FTP_TYPES.values}, if: :ftp_sync?
```
Comment (wvengen) [inline app/models/supplier.rb]: spaces?

File: app/models/supplier.rb (hunk 1 above)
Comment (JuliusR) [reply app/models/supplier.rb]: I changed it to the string approach with d664a26e99ac139785366d982b3d9b9de56faf5b. Let me rebase the whole branch before we merge.

File: app/models/supplier.rb (hunk 2 above)
Comment (JuliusR) [reply app/models/supplier.rb]: Added spaces in d664a26e99ac139785366d982b3d9b9de56faf5b .

**Comment by JuliusR** [comment]
> And I think it would be useful to allow specifying a file format, giving more flexibility. But then we'd also need to configure how to find the files on FTP (I do see `ftp_regexp`, sounds like a nice idea).

Yes, the `ftp_regexp` selects those files that should be downloaded from the FTP server. Obviously, different formats require different files. In addition, I can imagine that different suppliers are using different naming conventions. If I understand it correctly then the foodsoft format can even be used with non-csv files.

The `ftp_regexp` seemed like the easiest solution (i.e., let the sharedlists admin deal with that issue).

> Maybe we can choose a file format here, and let each file format specify a default ftp regexp? Just an idea.

Yes, you are right: The sharedlists application could suggest `ftp_type`-specific defaults. If those are not working then there could be a customization option. However, since we are dealing with an admin interface I prefer a less complex solution over a more user-friendly solution. How do you like the recently added [`ftp_regexp_hint`](https://github.com/foodcoops/sharedlists/pull/17/commits/d664a26e99ac139785366d982b3d9b9de56faf5b#diff-37c9b835927e3c363588f172de9bb6faR50)?

BTW: Currently, there is no validation done on `ftp_regexp`. I think that is alright since sharedlists admins should be trustworthy.

**Comment by JuliusR** [comment]
I am ready to rebase and combine the two database migrations into one if that is preferable. I am just waiting for feedback on the regexp issues.

Take your time though. I just want to avoid that we are wating for each other to proceed.

**Comment by JuliusR** [comment]
@wvengen, @all: I would like to merge this soon (because one of our suppliers is about to publish foodsoft-CSV files on a regular basis). If you do not complain then I will close this and reopen a rebased PR for merging.

**Pull request closed without merging**